src/optimizer/cardinality_estimator.cpp

```diff
@@ -0,0 +1,569 @@
+		relations_to_tdoms.at(set_i).equivalent_relations.insert(filter_info->right_binding);
+		relations_to_tdoms.at(set_i).filters.push_back(filter_info);
+	} else if (matching_equivalent_sets.empty()) {
+		column_binding_set_t tmp;
+		tmp.insert(filter_info->left_binding);
+		tmp.insert(filter_info->right_binding);
+		relations_to_tdoms.emplace_back(RelationsToTDom(tmp));
+		relations_to_tdoms.back().filters.push_back(filter_info);
+	}
+}
+
+void CardinalityEstimator::AddRelationToColumnMapping(ColumnBinding key, ColumnBinding value) {
+	relation_column_to_original_column[key] = value;
+}
+
+void CardinalityEstimator::CopyRelationMap(column_binding_map_t<ColumnBinding> &child_binding_map) {
+	for (auto &binding_map : relation_column_to_original_column) {
+		child_binding_map[binding_map.first] = binding_map.second;
+	}
+}
+
+void CardinalityEstimator::AddColumnToRelationMap(idx_t table_index, idx_t column_index) {
+	relation_attributes[table_index].columns.insert(column_index);
+}
+
+void CardinalityEstimator::InitEquivalentRelations(vector<unique_ptr<FilterInfo>> *filter_infos) {
+	// For each filter, we fill keep track of the index of the equivalent relation set
+	// the left and right relation needs to be added to.
+	vector<idx_t> matching_equivalent_sets;
+	for (auto &filter : *filter_infos) {
+		matching_equivalent_sets.clear();
+
+		if (SingleColumnFilter(filter.get())) {
+			continue;
+		}
+		D_ASSERT(filter->left_set->count >= 1);
+		D_ASSERT(filter->right_set->count >= 1);
+
+		matching_equivalent_sets = DetermineMatchingEquivalentSets(filter.get());
+		AddToEquivalenceSets(filter.get(), matching_equivalent_sets);
+	}
+}
+
+void CardinalityEstimator::VerifySymmetry(JoinNode *result, JoinNode *entry) {
+	if (result->GetCardinality() != entry->GetCardinality()) {
+		// Currently it's possible that some entries are cartesian joins.
+		// When this is the case, you don't always have symmetry, but
+		// if the cost of the result is less, then just assure the cardinality
+		// is also less, then you have the same effect of symmetry.
+		D_ASSERT(ceil(result->GetCardinality()) <= ceil(entry->GetCardinality()) ||
+		         floor(result->GetCardinality()) <= floor(entry->GetCardinality()));
+	}
+}
+
+void CardinalityEstimator::InitTotalDomains() {
+	auto remove_start = std::remove_if(relations_to_tdoms.begin(), relations_to_tdoms.end(),
+	                                   [](RelationsToTDom &r_2_tdom) { return r_2_tdom.equivalent_relations.empty(); });
+	relations_to_tdoms.erase(remove_start, relations_to_tdoms.end());
+}
+
+double CardinalityEstimator::ComputeCost(JoinNode *left, JoinNode *right, double expected_cardinality) {
+	double cost = expected_cardinality + left->GetCost() + right->GetCost();
+	return cost;
+}
+
+double CardinalityEstimator::EstimateCrossProduct(const JoinNode *left, const JoinNode *right) {
+	// need to explicity use double here, otherwise auto converts it to an int, then
```

Comment:
Nitpick: this can be simplified to:
```c++
return left->GetCardinality() >= (NumericLimits<double>::Maximum() / right->GetCardinality()) ?
       NumericLimits<double>::Maximum() : left->GetCardinality() * right->GetCardinality()
```